plan/exhaust_physical_plans.go

```diff
@@ -540,36 +540,46 @@ func (p *LogicalJoin) buildRangeForIndexJoin(indexInfo *model.IndexInfo, innerPl
 		}
 	}
 
-	return ranges, remained, keyOff2IdxOff
+	return ranges, append(remained, moreRemained...), keyOff2IdxOff
 }
 
 func (p *LogicalJoin) buildFakeEqCondsForIndexJoin(keys, idxCols []*expression.Column, colLengths []int,
-	innerPlan *DataSource) (accesses, eqConds []expression.Expression, keyOff2IdxOff []int) {
+	innerFilters []expression.Expression) (accesses, eqConds, remained []expression.Expression, keyOff2IdxOff []int) {
 	// Check whether all join keys match one column from index.
 	keyOff2IdxOff = joinKeysMatchIndex(keys, idxCols, colLengths)
 	if keyOff2IdxOff == nil {
-		return nil, nil, nil
+		return nil, nil, nil, nil
 	}
 
-	// After predicate push down, the one side conditions of join must be the conditions that cannot be pushed down and
-	// cannot calculate range either. So we only need the innerPlan.pushedDownConds and the eq conditions that we generate.
-	// TODO: There may be a selection that block the index join.
-	conds := make([]expression.Expression, 0, len(keys)+len(innerPlan.pushedDownConds))
+	usableKeys := make([]*expression.Column, 0, len(keys))
+
+	conds := make([]expression.Expression, 0, len(keys)+len(innerFilters))
 	eqConds = make([]expression.Expression, 0, len(keys))
-	// Construct a fake equal expression for calculating the range.
+	// Construct a fake equal expression for every join key for calculating the range.
 	for i, key := range keys {
 		if keyOff2IdxOff[i] < 0 {
 			continue
 		}
+		usableKeys = append(usableKeys, key)
 		// Int datum 1 can convert to all column's type(numeric type, string type, json, time type, enum, set) safely.
 		fakeConstant := &expression.Constant{Value: types.NewIntDatum(1), RetType: key.GetType()}
 		eqFunc := expression.NewFunctionInternal(p.ctx, ast.EQ, types.NewFieldType(mysql.TypeTiny), key, fakeConstant)
 		conds = append(conds, eqFunc)
 		eqConds = append(eqConds, eqFunc)
 	}
 
-	conds = append(conds, innerPlan.pushedDownConds...)
-	return conds, eqConds, keyOff2IdxOff
+	// Look into every `innerFilter`, if it contains join keys' column, put this filter into `remained` part directly.
+	remained = make([]expression.Expression, 0, len(innerFilters))
+	for _, filter := range innerFilters {
+		affectedCols := expression.ExtractColumns(filter)
+		if expression.ColumnSliceIsIntersect(affectedCols, usableKeys) {
+			remained = append(remained, filter)
+			continue
+		}
+		conds = append(conds, filter)
```

Comment:
Since this filter doesn't contain the column of join keys, can we just move all the filters in `innerFilters` to `remained`?

Reply:
t1.a=t2.a and t1.b=1 index is (a, b).

Reply:
got it.